kroxylicious/src/main/java/io/kroxylicious/proxy/filter/NetFilter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter;
+
+import java.net.SocketAddress;
+
+/**
+ * Abstracts some policy/logic for now an upstream connection for a given client connection
+ * is made.
+ */
+public interface NetFilter {
+
+    /**
+     * Determine the upstream cluster to connect to based on the information
+     * provided by the given {@code context},
+     * by invoking {@link NetFilterContext#connect(String, int, KrpcFilter[])}.
+     * @param context The context.
+     */
+    void upstreamBroker(NetFilterContext context);
```

Comment:
I accept that `upstreamBroker` is a poor name, but I don't think `connect` is quite right either. This method is where the proxy is asking the filter/plugin "what should I connect to?". It's up to the filter to initiate the connect by invoking `connect()` on the `NetFilterContext` passed in as an argument. What about:

* `NetFilter.upstreamBroker` → `NetFilter.selectBroker`
* `NetFilterContext.connect` → `NetFilterContext.initiateBrokerConnection`

Even then, I'm not super-convinced about using the word "broker". It assumes that the upstream server has the broker role. Maybe one day the proxy supports proxying servers which are Kraft controllers only. So maybe

* `NetFilter.upstreamBroker` → `NetFilter.selectServer`
* `NetFilterContext.connect` → `NetFilterContext.initiateConnection` (no server, because what else would you be connecting to? It can only be a server in the TCP sense).